keps/sig-cli/3895-kubectl-delete-interactivity/README.md

```diff
@@ -0,0 +1,926 @@
+Deployment/dockerd2
+Do you want to continue? (y/n): y
+deployment.apps "dockerd" deleted
+deployment.apps "dockerd2" deleted
+```
+
+For the alpha phase, this flag will be hidden behind `KUBECTL_INTERACTIVE_DELETE` environment variable and
+can only be use when this `KUBECTL_INTERACTIVE_DELETE=true` environment variable is exported.
+
+### Implementation Details/Design
+
+One of major advantage of this feature is the ability to preview all resources before deletion. That means
+we need to traverse all resources twice. First one is for display, and the other one is for deletion.
+However, that comes along with a technical limitation that `Visit` function does not support traversing twice
+as stated in [here](https://github.com/kubernetes/kubernetes/blob/6d7a9048b67d81e57923892650a23636f1afba42/staging/src/k8s.io/cli-runtime/pkg/resource/result.go#L91-L94).
+But we also cannot use `Infos` function as suggested in the link above, because that unties the builder configuration
+and deletion logic.
+
+In order to overcome this limitation, this feature will use separate builder to show resources as preview. Current builder
+will continue being used as deletion traverse, other builder will be used for preview. Builders must match in terms of 
+configurations to prevent such discrepancies and required tests will be added to assure that configurations are same or not.
+Moreover, deletion will only be performed to the resources whose `uidMap` match the ones in the preview list.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+As a user, I want to have an option via a flag to list all the resources being deleted as preview and
+proceed to actual deletion only after explicit confirmation.
+
+#### Story 1
+As filed in this issue https://github.com/kubernetes/kubectl/issues/1365;
+user tries to delete all persistentvolumeclaims in a namespace;
+```sh
+$ kubectl delete --all pvc -n yournamespace
+```
+
+However, mistyping `pv` instead of `pvc` deletes all persistentvolumes in the cluster, 
+since persistentvolumes are non-namespaced resources.
+
+#### Story 2
+As filed in this issue https://github.com/kubernetes/kubectl/issues/696;
+user executes `kubectl delete --all namespaces` without enough knowledge about the
+consequences of this command.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+- In terms of backwards compatibility, this flag's default will always be false.
+- If user cancels deletion, exit code will be 0.
+- This flag will not be used with `--raw` flag. 
```

Comment:
How does this change behave in regards to label based deletion, especially if the set of objects matched by the label changes?

Reply:
This problem is also true for default delete behavior(although agree that this is much more visible with this interactivity flag). We'll delete the resources whose are listed as preview and user confirms. If set of objects belonging to the some label is changed, user should execute `kubectl delete` again.

Reply:
Include the intended behavior in the KEP.

Reply:
Updated. Does it make sense?. Thanks.